book/src/rfcs/benchmarking.md

```diff
@@ -0,0 +1,109 @@
+# Benchmarking RFC
+
+This document proposes a high level approach to benchmarking the crypto stack, mostly kimchi, snarky and pickles. The main concern is to set up an infrastructure where it is possible to integrate different kinds of benchmarks with minimal effort. While the specific benchmarks are not an immediate concern, a few benchmarks of interest will be mentioned later.
+
+## Types of benchmarks
+
+There are different metrics that we can collect, and different ways of presenting them depending of who is interested in them.
+
+### Audience
+
+We can differentiate to groups of people interested in benchmarks:
+
+- Developers working in the crypto stack, mostly our crypto team, where the main interest is to avoid regressions or make performance improvements.
+- People working outside of it, mostly users of our libraries, but could also include decision makers comparing other options or our marketing team, here the interest is more about more general metrics that can be compared with the competition or used to learn how to most efficiently use our technology.
+
+Must be mentioned that some overlap should be expected, some benchmarks may be useful for anyone while other just to some.
+
+### Metrics
+
+There are two main metrics, performance and memory, and a smaller one hardware utilization.
+
+#### Performance
+
+Basically execution time, this may be the hardest to measure, because the environment can have a considerable effect over the results, some high quality machine would be required, ideally a fully dedicated one to minimize noise, there is a way of measuring that overcomes noise problems that will be mentioned.
+It can be subdivided it 2 other metrics or alternative approaches:
+
+- Execution time: The most simple, run the program in question several times and take the average, it may become quite expensive if the program measured takes some time because that time will be multiplied by the number of samples, in those cases it can be made faster sacrificing precision. This benchmark is one of the most interesting for external users, and one of the main comparison point for evaluating the competition. It can also serve for internal developer as a regression test or to verify performance improvements.
+
+- CPU instructions: This method instead of recording execution time counts the number of cpu instructions used, not very useful for external developer given that it doesn't translate directly to time. It is mostly useful to internal developer as an alternative to execution time, it is faster given that you only have to run it once to collect the information and can be run on cheap machines without significative noise given that execution time doesn't affect the number of instructions. Additionally the tools used to measure it also collect some other metrics that can be useful to aide optimization like cache misses and branch mispredictions.
```

Comment:
```suggestion
: CPU instructions
Instead of recording execution time, this method counts the number of cpu instructions used in a program execution, not very useful for external developer given that it doesn't translate directly to time. It is mostly useful to our engineering teams as an alternative to execution time, it is faster given that you only have to run it once to collect the information, and can be run on cheap machines without significative noise given that execution time doesn't affect the number of instructions. Additionally, the tools used to measure it also collect some other metrics that can be useful to aide optimization like cache misses and branch mispredictions.
```